src/dune_engine/build_system.ml

```diff
@@ -1399,9 +1410,8 @@ end = struct
       (let open Fiber.O in
       let build_deps deps = Memo.Build.run (build_deps deps) in
       report_evaluated_rule t;
-      Fs.mkdir_p dir;
+      let* () = Memo.Build.run (Fs.mkdir_p dir) in
```

Comment:
At first, I thought that it was a bit sad to introduce an extra layer of complexity for something as simple as `mkdir` but then I realised that actually we might want to track deletions of directories created by `mkdir`, in which case we would want to rerun `mkdir` only when necessary. So, in the end, it may be worth it!